Add Check Support for Java 21 Pattern Matching for Switch Syntax: Indentation

child of #14961 
I have read check documentation: https://checkstyle.org/checks/misc/indentation.html#Indentation
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> cat src/Test.java                                                 
record Point(int x, int y) { }
record square(Point upperLeft, Point lowerRight) { }

public class Test {
    public void test() {
        Point p = new Point(1, 2);
        switch (p) {
            case Point(int x, int y)
            when x > 0 && y > 0 -> System.out.println(x + y); // violation
            case Point(int x, int y) -> System.out.println("error");
        }
    }
}
PS D:\CS\test> cat config.xml                                                    
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="Indentation">
    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:13: 'case' child has incorrect indentation level 12, expected level should be 16. [Indentation]
Audit done.
Checkstyle ends with 1 errors.
```
---

**Describe what you expect in detail.**

I consider this a false positive. I want no violation on `when`

---

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheckTest.java
@@ -3019,6 +3019,34 @@ public class IndentationCheckTest extends AbstractModuleTestSupport {
         verifyWarns(checkConfig, fileName, expected);
     }
 
+    @Test
+    public void testIndentationPatternMatchingForSwitch()
+            throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(IndentationCheck.class);
+        checkConfig.addProperty("forceStrictCondition", "true");
+        checkConfig.addProperty("tabWidth", "4");
+        checkConfig.addProperty("basicOffset", "4");
+        checkConfig.addProperty("braceAdjustment", "0");
+        checkConfig.addProperty("caseIndent", "4");
+        checkConfig.addProperty("throwsIndent", "8");
+
+        final String fileName = getNonCompilablePath(
+                "InputIndentationPatternMatchingForSwitch.java");
+        final String[] expected = {
+            "21:13: " + getCheckMessage(MSG_CHILD_ERROR, "case", 12, 16),
+            "54:13: " + getCheckMessage(MSG_CHILD_ERROR, "case", 12, 16),
+            "69:13: " + getCheckMessage(MSG_CHILD_ERROR, "case", 12, 16),
+            "70:13: " + getCheckMessage(MSG_CHILD_ERROR, "case", 12, 16),
+            "75:5: " + getCheckMessage(MSG_CHILD_ERROR, "case", 4, 16),
+            "76:5: " + getCheckMessage(MSG_CHILD_ERROR, "case", 4, 16),
+            "87:1: " + getCheckMessage(MSG_CHILD_ERROR, "case", 0, 16),
+            "88:1: " + getCheckMessage(MSG_CHILD_ERROR, "case", 0, 16),
+            "89:1: " + getCheckMessage(MSG_CHILD_ERROR, "case", 0, 16),
+            "90:1: " + getCheckMessage(MSG_ERROR, "lambda", 0, 16),
+        };
+        verifyWarns(checkConfig, fileName, expected);
+    }
+
     private static final class IndentAudit implements AuditListener {
 
         private final IndentComment[] comments;

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationPatternMatchingForSwitch.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationPatternMatchingForSwitch.java
@@ -1,0 +1,95 @@
+//non-compiled with javac: Compilable with Java21        //indent:0 exp:0
+package com.puppycrawl.tools.checkstyle.checks.indentation.indentation;//indent:0 exp:0
+//indent:0 exp:0
+//indent:0 exp:0
+/* Config:                                                                  //indent:0 exp:0
+ * basicOffset = 4                                                          //indent:1 exp:1
+ * braceAdjustment = 0                                                      //indent:1 exp:1
+ * caseIndent = 4                                                           //indent:1 exp:1
+ * tabWidth = 4                                                             //indent:1 exp:1
+ * throwsIndent = 8                                                         //indent:1 exp:1
+ * forceStrictCondition = true                                             //indent:1 exp:1
+ */                                                                         //indent:1 exp:1
+public class InputIndentationPatternMatchingForSwitch {                     //indent:0 exp:0
+    record Point(int x, int y) { }                                           //indent:4 exp:4
+    record square(Point upperLeft, Point lowerRight) { }                    //indent:4 exp:4
+
+    public void test1() {   //indent:4 exp:4
+        Point p = new Point(1, 2); //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case Point(int x, int y) //indent:12 exp:12
+            when x > 0 && y > 0 -> System.out.println(x + y); //indent:12 exp:16 warn
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case Point(int x, int y) //indent:12 exp:12
+                when x > 0 && y > 0 -> System.out.println(x + y); //indent:16 exp:16
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+    } //indent:4 exp:4
+
+    public void test2() { //indent:4 exp:4
+        Point p = new Point(1, 2); //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case    //indent:12 exp:12
+                    Point(int x, int y) //indent:20 exp:20
+                when x > 0 && y > 0 -> System.out.println(x + y); //indent:16 exp:16
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case    //indent:12 exp:12
+                    Point(int x, int y) //indent:20 exp:20
+                    when x > 0 && y > 0 -> System.out.println(x + y); //indent:20 exp:20
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case    //indent:12 exp:12
+                    Point(int x, int y) //indent:20 exp:20
+                            when x > 0 && y > 0 -> System.out.println(x + y); //indent:28 exp:28
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case    //indent:12 exp:12
+                    Point(int x, int y) //indent:20 exp:20
+            when x > 0 && y > 0 -> System.out.println(x + y); //indent:12 exp:16 warn
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+    } //indent:4 exp:4
+
+    public void test3() { //indent:4 exp:4
+        Point p = new Point(1, 2); //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case //indent:12 exp:12
+                Point(int x, int y) //indent:16 exp:16
+                when x > 0 && y > 0 -> System.out.println(x + y); //indent:16 exp:16
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case //indent:12 exp:12
+            Point(int x, int y) //indent:12 exp:16 warn
+            when x > 0 && y > 0 -> System.out.println(x + y); //indent:12 exp:16 warn
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case //indent:12 exp:12
+    Point(int x, int y) //indent:4 exp:16 warn
+    when x > 0 && y > 0 -> System.out.println(x + y); //indent:4 exp:16 warn
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case //indent:12 exp:12
+                Point(int x, int y) //indent:16 exp:16
+                            when x > 0 && y > 0 -> System.out.println(x + y); //indent:28 exp:28
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+        switch (p) { //indent:8 exp:8
+            case    //indent:12 exp:12
+Point(int x, int y) //indent:0 exp:16 warn
+when          //indent:0 exp:16 warn
+x > 0 && y > 0          //indent:0 exp:16 warn
+->        //indent:0 exp:16 warn
+System.out.println(x + y); //indent:0 exp:0
+            case Point(int x, int y) -> System.out.println("error"); //indent:12 exp:12
+        } //indent:8 exp:8
+    } //indent:4 exp:4
+}//indent:0 exp:0

